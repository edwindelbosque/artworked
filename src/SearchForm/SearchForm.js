import React, { Component } from 'react';
import './SearchForm.scss';

class SearchForm extends Component {
	constructor() {
		super();
		this.state = {
			term: '',
			type: ''
		};
	}

	render() {
		return <h2>SearchForm</h2>;
	}
}

export default SearchForm;
