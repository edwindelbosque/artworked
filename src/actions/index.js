export const setResults = results => ({
	type: 'SET_RESULTS',
	results
});

export const toggleLoading = boolean => ({
	type: 'TOGGLE_LOADING',
	boolean
});

export const addFavorite = item => ({
	type: 'ADD_FAVORITE',
	item
});

export const removeFavorite = id => ({
	type: 'REMOVE_FAVORITE',
	id
});
