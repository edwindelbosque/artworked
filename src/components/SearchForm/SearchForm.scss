.SearchForm {
	width: 580px;
	padding: 20px 30px 10px 30px;
	display: flex;
	justify-content: space-between;
	align-items: flex-start;
	border: 4px solid rgb(130, 238, 202);
	border-top: none;
	border-radius: 0 0 4px 4px;

	div {
		display: flex;
		align-items: flex-end;
		flex-direction: column;
	}

	input {
		font-size: 14px;
		font-weight: 600;
		padding: 8px 10px;
		height: 30px;
		width: 190px;
		background: rgb(230, 230, 230);
		border: none;
		margin-bottom: 10px;
		border-radius: 0px;
	}

	button {
		font-size: 14px;
		font-weight: 600;
		padding: 7px 20px;
		color: white;
		background: #fe5752;
		border: none;
		border-radius: 0px;
		transition: 0.2s;

		&:hover {
			background: #ff6536;
			transition: 0.2s;
		}

		&:focus {
			outline: none;
		}

		&:active {
			background: #be4a27;
		}
	}

	select {
		font-size: 14px;
		font-weight: 600;
		padding: 8px 10px;
		height: 30px;
		background: rgb(230, 230, 230);
		width: 143px;
		border: none;
		border-radius: 2px;
	}
}
