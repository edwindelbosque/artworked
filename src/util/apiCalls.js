const baseSearchUrl = 'https://itunes.apple.com/search?term=';
const baseLookupUrl = 'https://itunes.apple.com/lookup?id=';

export const getData = async (search, type) => {
	try {
		const response = await fetch(`${baseSearchUrl}${search}&entity=${type}`);
		const data = await response.json();
		console.log(data);
		return data;
	} catch (error) {
		throw new Error('Failed to get results');
	}
};

export const getAlbumTracks = async collectionId => {
	try {
		const response = await fetch(`${baseLookupUrl}${collectionId}&entity=song`);
		const albumTracks = await response.json();
		return albumTracks;
	} catch (error) {
		throw new Error('Failed to get album tracks');
	}
};
