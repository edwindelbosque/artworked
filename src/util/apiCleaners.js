export const cleanerHandler = (data, label, artist, term) => {
	const mutableData = [...data.results];

	if (label === 'Album') {
		const cleanAlbums = artist
			? cleanAlbum(filterArtist(artist, mutableData))
			: cleanAlbum(mutableData, term);
		return filterResults(cleanAlbums, term);
	}
	if (label === 'Single') {
		const cleanSingles = artist
			? cleanSingle(filterArtist(artist, mutableData))
			: cleanSingle(mutableData);
		return filterResults(cleanSingles, term);
	}
	if (label === 'Artist') {
		return cleanArtist(mutableData);
	}
	if (label === 'Movie') {
		const cleanMovies = cleanMovie(mutableData);
		return filterResults(cleanMovies, term);
	}
	if (label === 'App') {
		return cleanApp(mutableData.slice(0, 1));
	}
	if (label === 'Podcast') {
		return cleanPodcast(mutableData);
	}
	if (label === 'Music Video') {
		return cleanMusicVideo(mutableData);
	}
	if (label === 'TV Show') {
		return cleanTvShow(mutableData);
	}
	if (label === 'iBook') {
		return cleanBook(mutableData);
	}
	if (label === 'Audiobook') {
		return cleanAudiobook(mutableData);
	}
};

const filterResults = (cleanData, term) => {
	return cleanData.filter(result => {
		const upperTerm = term.toUpperCase().replace(/ /g, '');
		const upperName = result.name.toUpperCase().replace(/ /g, '');
		return upperName.includes(upperTerm);
	});
};

const filterArtist = (artist, data) => {
	return data.filter(result => {
		const flattedTerm = artist
			.toUpperCase()
			.replace(/ /g, '')
			.normalize('NFD')
			.replace(/[\u0300-\u036f]/g, '');
		const flattedArtistName = result.artistName
			.toUpperCase()
			.replace(/ /g, '')
			.normalize('NFD')
			.replace(/[\u0300-\u036f]/g, '');
		return flattedArtistName.includes(flattedTerm);
	});
};

const cleanArtwork = (artworkUrl, resolution) => {
	const splitUrl = artworkUrl.split('');
	splitUrl.splice(splitUrl.length - 13);
	const joinUrl = splitUrl.join('');
	const cleanUrl = `${joinUrl}${resolution}x${resolution}.jpg`;
	return cleanUrl;
};

const cleanAlbum = mutableData => {
	const filteredData = mutableData.filter(result => result.trackCount > 5);
	return filteredData.map(result => {
		const {
			artistId,
			artistName,
			artworkUrl100,
			collectionId,
			collectionName,
			releaseDate,
			trackCount
		} = result;
		const cleanedData = {
			artistId,
			artist: artistName,
			artwork: cleanArtwork(artworkUrl100, 580),
			hqArtwork: cleanArtwork(artworkUrl100, 10000),
			albumId: collectionId,
			name: collectionName,
			releaseYear: parseInt(releaseDate.slice(0, 4)),
			trackCount
		};
		return cleanedData;
	});
};

const cleanSingle = mutableData => {
	const filteredData = mutableData.filter(result => result.trackCount <= 5);
	return filteredData.map(result => {
		const {
			artistId,
			artistName,
			artworkUrl100,
			collectionId,
			collectionName,
			releaseDate,
			trackCount
		} = result;
		const cleanedData = {
			artistId,
			artist: artistName,
			artwork: cleanArtwork(artworkUrl100, 580),
			hqArtwork: cleanArtwork(artworkUrl100, 10000),
			albumId: collectionId,
			name: collectionName,
			releaseYear: parseInt(releaseDate.slice(0, 4)),
			trackCount
		};
		return cleanedData;
	});
};

const cleanArtist = mutableData => {
	return mutableData.map(result => {
		const {
			artistId,
			artistName,
			artworkUrl100,
			collectionId,
			collectionName,
			releaseDate,
			trackCount
		} = result;
		const cleanedData = {
			artistId,
			artist: artistName,
			artwork: cleanArtwork(artworkUrl100, 580),
			hqArtwork: cleanArtwork(artworkUrl100, 10000),
			albumId: collectionId,
			name: collectionName,
			releaseYear: parseInt(releaseDate.slice(0, 4)),
			trackCount
		};
		return cleanedData;
	});
};

const cleanMovie = mutableData => {
	return mutableData.map(result => {
		const { artworkUrl100, trackName, releaseDate } = result;

		const cleanedData = {
			artwork: cleanArtwork(artworkUrl100, 580),
			hqArtwork: cleanArtwork(artworkUrl100, 10000),
			name: trackName,
			releaseYear: parseInt(releaseDate.slice(0, 4))
		};
		return cleanedData;
	});
};

const cleanPodcast = mutableData => {
	return mutableData.map(result => {
		const { artworkUrl100, trackName, releaseDate } = result;

		const cleanedData = {
			artwork: cleanArtwork(artworkUrl100, 580),
			hqArtwork: cleanArtwork(artworkUrl100, 10000),
			name: trackName,
			releaseYear: parseInt(releaseDate.slice(0, 4))
		};
		return cleanedData;
	});
};

const cleanMusicVideo = mutableData => {
	return mutableData.map(result => {
		const {
			artworkUrl100,
			trackName,
			releaseDate,
			previewUrl,
			artistName
		} = result;

		const cleanedData = {
			artist: artistName,
			artwork: cleanArtwork(artworkUrl100, 580),
			hqArtwork: cleanArtwork(artworkUrl100, 10000),
			previewUrl,
			name: trackName,
			releaseYear: parseInt(releaseDate.slice(0, 4))
		};
		return cleanedData;
	});
};

const cleanTvShow = mutableData => {
	return mutableData.map(result => {
		const { artworkUrl100, releaseDate, collectionName } = result;

		const cleanedData = {
			artwork: cleanArtwork(artworkUrl100, 580),
			hqArtwork: cleanArtwork(artworkUrl100, 10000),
			name: collectionName,
			releaseYear: parseInt(releaseDate.slice(0, 4))
		};
		return cleanedData;
	});
};

const cleanBook = mutableData => {
	return mutableData.map(result => {
		const { artworkUrl100, trackName, releaseDate, artistName } = result;

		const cleanedData = {
			artist: artistName,
			artwork: cleanArtwork(artworkUrl100, 580),
			hqArtwork: cleanArtwork(artworkUrl100, 10000),
			name: trackName,
			releaseYear: parseInt(releaseDate.slice(0, 4))
		};
		return cleanedData;
	});
};

const cleanAudiobook = mutableData => {
	return mutableData.map(result => {
		const { artworkUrl100, collectionName, releaseDate, artistName } = result;

		const cleanedData = {
			artist: artistName,
			artwork: cleanArtwork(artworkUrl100, 580),
			hqArtwork: cleanArtwork(artworkUrl100, 10000),
			name: collectionName,
			releaseYear: parseInt(releaseDate.slice(0, 4))
		};
		return cleanedData;
	});
};

const cleanApp = mutableData => {
	return mutableData.map(result => {
		const { artworkUrl100, trackName, releaseDate } = result;

		const cleanedData = {
			artwork: cleanArtwork(artworkUrl100, 580),
			hqArtwork: cleanArtwork(artworkUrl100, 10000),
			name: trackName,
			releaseYear: parseInt(releaseDate.slice(0, 4))
		};
		return cleanedData;
	});
};
