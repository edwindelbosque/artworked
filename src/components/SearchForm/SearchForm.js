import React, { Component } from 'react';
import './SearchForm.scss';
import { getData } from '../../util/apiCalls';
import { cleanerHandler } from '../../util/apiCleaners';

class SearchForm extends Component {
	constructor() {
		super();
		this.state = {
			type: 'album',
			term: '',
			artist: '',
			label: 'Album',
			placeholder: 'ex. Pop2'
		};
	}

	handleChange = e => {
		this.setState({
			[e.target.name]: e.target.value
		});

		if (e.target.name === 'type') {
			const value = JSON.parse(e.target.value);
			this.setState({
				type: value[0],
				label: value[1],
				placeholder: value[2]
			});
		}
	};

	handleClick = async () => {
		const { term, type, label } = this.state;
		const data = await getData(term, type);
		cleanerHandler(data, label);
	};

	render() {
		const { term, artist, label, placeholder } = this.state;

		const artistInput = (
			<div>
				<label>Artist Name</label>
				<input
					type='text'
					name='artist'
					placeholder='(higher accuracy)'
					value={artist}
					onChange={e => this.handleChange(e)}
				/>
			</div>
		);

		return (
			<nav className='SearchForm'>
				<div>
					<label>Look for </label>
					<select name='type' onChange={e => this.handleChange(e)}>
						<option value='["album", "Album", "ex. Pop2"]'>Album</option>
						<option value='["album", "Single", "ex. West Coast"]'>
							Single EP
						</option>
						<option value='["album", "Artist", "ex. Joseph Salvat"]'>
							Artist
						</option>
						<option value='["tvSeason", "TV Show", "ex. The Office"]'>
							TV Show
						</option>
						<option value='["movie", "Movie", "ex. Joker"]'>Movie</option>
						<option value='["podcast", "Podcast", "ex. Reply All"]'>
							Podcast
						</option>
						<option value='["ebook", "iBook", "ex. The Martian"]'>iBook</option>
						<option value='["audiobook", "Audiobook", "ex. The Martian"]'>
							Audiobook
						</option>
						<option value='["software", "App", "ex. Instagram"]'>App</option>
						<option value='["musicVideo", "Music Video", "ex. Shades of Cool"]'>
							Music Video
						</option>
					</select>
				</div>
				<div>
					<label>{label} Name</label>
					<input
						type='text'
						name='term'
						placeholder={placeholder}
						value={term}
						onChange={e => this.handleChange(e)}
					/>
				</div>
				{label === 'Single' && artistInput}
				{label === 'Album' && artistInput}
				<button onClick={this.handleClick}>Find Artwork</button>
			</nav>
		);
	}
}

export default SearchForm;
