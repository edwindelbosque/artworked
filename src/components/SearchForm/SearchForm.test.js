import React from 'react';
import { SearchForm, mapDispatchToProps } from './SearchForm';
import { shallow } from 'enzyme';
import { getData } from '../../util/apiCalls';
import { setResults, toggleLoading } from '../../actions';

jest.mock('../../util/apiCalls');

describe('SearchForm', () => {
	let wrapper;
	const mockEventAlbum = { target: { name: 'term', value: 'Pop 2' } };
	const mockEventArtist = { target: { name: 'artist', value: 'Charli XCX' } };
	const mockEventSelect = {
		target: { name: 'type', value: '["album", "Album"]' }
	};

	beforeEach(() => {
		wrapper = shallow(<SearchForm />);
	});

	it('should match SearchForm snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('should update state when user types on inputs or changes select', () => {
		const expectedInputAlbum = 'Pop 2';
		const expectedInputArtist = 'Charli XCX';
		const expectedSelectTypeValue = 'album';
		const expectedSelectTypeLabel = 'Album';

		wrapper.instance().handleChange(mockEventAlbum);
		wrapper.instance().handleChange(mockEventArtist);
		wrapper.instance().handleChange(mockEventSelect);

		expect(wrapper.state('term')).toEqual(expectedInputAlbum);
		expect(wrapper.state('artist')).toEqual(expectedInputArtist);
		expect(wrapper.state('type')).toEqual(expectedSelectTypeValue);
		expect(wrapper.state('label')).toEqual(expectedSelectTypeLabel);
	});

	it('should invoke handleChange when user types or changes select', () => {
		wrapper.instance().handleChange = jest.fn();

		wrapper
			.find('input')
			.at(0)
			.simulate('change', mockEventAlbum);

		expect(wrapper.instance().handleChange).toHaveBeenCalled();
	});

	it('should invoke handleChange when user types or changes select', () => {
		wrapper.instance().handleChange = jest.fn();

		wrapper.find('select').simulate('change', mockEventSelect);

		expect(wrapper.instance().handleChange).toHaveBeenCalled();
	});

	it('should invoke handleChange when user types or changes select', () => {
		wrapper.instance().handleChange = jest.fn();

		wrapper.find('input').simulate('change', mockEventArtist);

		expect(wrapper.instance().handleChange).toHaveBeenCalled();
	});

	it('should invoke handleSubmit on submit', () => {
		wrapper.instance().handleSubmit = jest.fn();
		const mockEvent = {
			preventDefault: jest.fn()
		};

		wrapper.find('form').simulate('submit', mockEvent);

		expect(wrapper.instance().handleSubmit).toHaveBeenCalled();
	});

	it('should invoke handleClick on click', () => {
		wrapper.instance().handleClick = jest.fn();

		wrapper
			.find('input')
			.at(0)
			.simulate('change', mockEventAlbum);
		wrapper.find('button').simulate('click');

		expect(wrapper.instance().handleClick).toHaveBeenCalled();
	});

	it('should reset state when clearInputs is called', () => {
		const expected = { term: '', artist: '', label: 'Artist', type: 'album' };

		wrapper.instance().setState({
			term: 'Pop 2',
			artist: 'Charli XCX'
		});
		wrapper.instance().resetState();
		expect(wrapper.state()).toEqual(expected);
	});

	it('should call GetData fetch when handleClick is called', () => {
		const mockTerm = 'Pop 2';
		const mockType = 'album';
		wrapper.find('input').simulate('change', mockEventAlbum);
		wrapper.instance().handleClick();
		expect(getData).toHaveBeenCalledWith(mockTerm, mockType);
	});

	describe('mapDispatchToProps', () => {
		const results = [{}, {}, {}];
		const mockDispatch = jest.fn();

		it('should call dispatch with a setResults action when handleClick is called', () => {
			const actionToDispatch = setResults(results);

			const mappedProps = mapDispatchToProps(mockDispatch);
			mappedProps.setResults(results);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
		});

		it('should call dispatch with a toggleLoading true action when handleClick is called', () => {
			const actionToDispatch = toggleLoading(true);

			const mappedProps = mapDispatchToProps(mockDispatch);
			mappedProps.toggleLoading(true);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
		});

		it('should call dispatch with a toggleLoading false action when handleClick is called', () => {
			const actionToDispatch = toggleLoading(false);

			const mappedProps = mapDispatchToProps(mockDispatch);
			mappedProps.toggleLoading(false);

			expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
		});
	});
});
