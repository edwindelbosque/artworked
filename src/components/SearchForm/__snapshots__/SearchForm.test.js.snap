// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match SearchForm snapshot 1`] = `
<form
  className="SearchForm"
  onSubmit={[Function]}
>
  <div
    className="inputsSection"
  >
    <select
      name="type"
      onChange={[Function]}
    >
      <option
        value="[\\"album\\", \\"Artist\\"]"
      >
        Artist
      </option>
      <option
        value="[\\"album\\", \\"Album\\"]"
      >
        Album
      </option>
      <option
        value="[\\"album\\", \\"Single\\"]"
      >
        Single
      </option>
      <option
        value="[\\"tvSeason\\", \\"TV Show\\"]"
      >
        TV Show
      </option>
      <option
        value="[\\"movie\\", \\"Movie\\"]"
      >
        Movie
      </option>
      <option
        value="[\\"podcast\\", \\"Podcast\\"]"
      >
        Podcast
      </option>
      <option
        value="[\\"ebook\\", \\"iBook\\"]"
      >
        iBook
      </option>
      <option
        value="[\\"audiobook\\", \\"Audiobook\\"]"
      >
        Audiobook
      </option>
    </select>
    <input
      name="term"
      onChange={[Function]}
      placeholder="Enter Artist"
      type="text"
      value=""
    />
  </div>
  <button
    onClick={[Function]}
  >
    FIND ARTWORK
  </button>
</form>
`;
